baseapp/msg_service_router.go

```diff
@@ -0,0 +1,97 @@
+package baseapp
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gogo/protobuf/proto"
+
+	gogogrpc "github.com/gogo/protobuf/grpc"
+	"google.golang.org/grpc"
+
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// MsgServiceRouter routes fully-qualified Msg service methods to their handler.
+type MsgServiceRouter struct {
+	interfaceRegistry codectypes.InterfaceRegistry
+	routes            map[string]MsgServiceHandler
+}
+
+var _ gogogrpc.Server = &MsgServiceRouter{}
+
+// NewMsgServiceRouter creates a new MsgServiceRouter.
+func NewMsgServiceRouter() *MsgServiceRouter {
+	return &MsgServiceRouter{
+		routes: map[string]MsgServiceHandler{},
+	}
+}
+
+// MsgServiceHandler defines a function type which handles Msg service message.
+type MsgServiceHandler = func(ctx sdk.Context, req sdk.MsgRequest) (*sdk.Result, error)
+
+// Handler returns the MsgServiceHandler for a given query route path or nil
+// if not found.
+func (msr *MsgServiceRouter) Handler(methodName string) MsgServiceHandler {
+	handler, found := msr.routes[methodName]
+	if !found {
+		return nil
+	}
+
+	return handler
```

Comment:
Why not just return `h, _ := msr.routes[methodName]`? Unless a key would exist w/o a handler somehow.

Reply:
> Unless a key would exist w/o a handler somehow.

even if a key exists w/o a handler, then `h` will be `nil`, right? so we're still good